Save metrics

Zwicii opened the pull request:
## What?
 
- Added pickle file saving for metrics data in utils/metrics.py
- Enhanced plot_mc_curve to show best metric value and confidence level for each class
- Created Plotly-based visualization notebook for interactive analysis

## Why?

- Metrics were only available as static plots during validation
- No easy way to identify optimal confidence thresholds per class
- Need for flexible post-analysis and interactive visualization
 
## How?

1. Pickle saving: ap_per_class function now saves metrics.pkl with raw data
2. Best per-class: plot_mc_curve calculates and displays optimal values for each class
3. Plotly integration: Notebook for loading pickle files and creating interactive plots

## Backout plan

Revert the changes


## Testing
 
Metrics file included in validation output folder:
<img width="700" height="164" alt="image" src="https://github.com/user-attachments/assets/220093e7-84ed-493c-890f-0ef0fd8f85a1" />


Best Per Class Info is available
<img width="963" height="642" alt="image" src="https://github.com/user-attachments/assets/0c335b99-581d-45a4-a916-1a3673ffc2c1" />


Custom pretty Plots generated with metrics.ipynb
<img width="879" height="455" alt="image" src="https://github.com/user-attachments/assets/203ee93e-182f-4788-a691-0948fa6378f8" />

Zwicii committed:
save metrics to pkl file

Zwicii committed:
add best score info for each class

Zwicii committed:
Add metrics notebook for visualizing pr, f1, p and r curves

Zwicii committed:
clean metrics notebook

kevinconka commented on utils/metrics.py:
```diff
@@ -365,7 +384,12 @@ def plot_mc_curve(px, py, save_dir=Path("mc_curve.png"), names=(), xlabel="Confi
         for i, y in enumerate(py):
-            ax.plot(px, y, linewidth=1, label=f"{names[i]}")  # plot(confidence, metric)
+
+            y_smooth = smooth(y, 0.05)
```
why are we smoothing it?

Zwicii replied:
So the best confidence for all classes is determined on the smoothed curve - so i thought we should to the same for the per class best confidences

https://github.com/SEA-AI/yolov5/blob/92b338af969f47aa3d0d859550803179927b5fff/utils/metrics.py#L396-L397
and determining the best confidences based on smoothed curves make it more robust to outlier in the data 

but its just my opinion :) i can of course change it

kevinconka commented on utils/metrics.py:
```diff
@@ -87,6 +87,25 @@ def ap_per_class(tp, conf, pred_cls, target_cls, plot=False, save_dir=".", names
+
+    # Save metrics to pickle
+    if save_dir:
+        import pickle
```
why did we choose to save it as pickle? Are the any advantages? I am concerned that pickle files are:

- not human readable
- not interoperable

Zwicii replied:
So first i had json --> was a mess to read because of the tabular data
So then i tought --> tablular data and human readable --> CVS but putting all the different infos in a CVS file felt a bit messy and multiple csv files seemed an overkill

then i thought ok i want to use the data 1:1  - just in another python script --> so i put them into a pkl file because reading and writing is super easy. And we could visualize in python.

but its just my opinion :) i can of course change it

kevinconka reviewed (commented):
Thank you so much for the color-blind palette!

kevinconka committed:
Refactor metrics saving functionality and add dashboard generation

- Consolidated metrics saving into dedicated functions: `save_metrics_to_pickle` and `save_metrics_dashboard`.
- Added error handling for dashboard saving to prevent import errors.
- Enhanced dashboard generation with Plotly for visualizing precision-recall and other metrics curves interactively in ah html

kevinconka reviewed (dismissed):
<img src="https://media0.giphy.com/media/v1.Y2lkPWJkM2VhNTdlenc4ZXQzeXlmcWlmeHFxdjE1YXRjNGw1Zzc3OWo4a2s0aHN4ZWFmMSZlcD12MV9naWZzX3NlYXJjaCZjdD1n/LedVYzRx24nkI/giphy-downsized-medium.gif"/>

kevinconka committed:
removed one-line function, bad claude

kevinconka committed:
Add logging for missing Plotly in metrics dashboard

- Introduced a warning log when Plotly is not installed, preventing silent failures during metrics dashboard generation.

Zwicii committed:
lazy import to resolve circular import

Zwicii committed:
add warning sign to plottly warning message
